File: frontend/src/types/highTrends.ts
/** 高潜趋势接口保留来源和统计口径，供人工复核，不把提及人数解释为偏好率。 */
export interface HighTrendRequest {
  dataset?: 'original' | 'article_table_2_selected_5'
  all_dates?: boolean
  start_date: string
  end_date: string
  model_id: string
  max_calls: number
  prompt?: string
  user_scope?: 'auto' | 'all' | 'first'
  user_limit?: number | null
}

/** 数量筛选在整理资料前生效，来源说明保留到历史任务。 */
export interface HighTrendScope {
  user_limit: number | null
  origin: 'option' | 'prompt' | 'default'
  label: string
  note: string
}

export interface HighTrendCatalog {
  trend_count: number
  user_count: number
  min_date: string | null
  max_date: string | null
  undated_count: number
}

export interface HighTrendPreview {
  scope: HighTrendScope
  counts: { source_users: number; selected_trends: number; selected_users: number; users_with_text: number; user_records: number }
  plan: { planned_calls: number; map_jobs: number; input_source_chars: number }
  warnings?: string[]
}

export interface TrendSource {
  short_id: string
  record_id?: string
  kind: string
  user_id?: string | number | null
  title?: string | null
  source_path?: string | null
  json_pointer?: string | null
  excerpt?: string | null
  question?: string | null
}

export interface TrendImage {
  image_id?: string | null
  code?: string | null
  source_record_id?: string
  role?: string
  kind?: string
  file_exists: boolean
  url?: string | null
}

export interface HighTrendCard {
  clustering_labels?: string[]
  id: string
  title: string
  description: string
  source_records: TrendSource[]
  /** 背景资料仅供追溯，不参与主卡图片关联；旧结果可能没有这些字段。 */
  background_source_records?: TrendSource[]
  background_source_ids?: string[]
  image_refs: TrendImage[]
  mention_statistics: { unique_mentioned_users: number; population_count: number; note: string }
  trend_coverage_note?: string
  reason?: string
}

export interface HighTrendResult {
  trend_categories?: { clustering_label: string; article_count: number; trend_ids: string[] }[]
  user_images?: UserResearchImage[]
  user_image_filter?: 'like_or_enjoy'
  user_images_scope_note?: string
  trends: HighTrendCard[]
  user_research_gaps: { directions: HighTrendCard[]; scope_note?: string; count_note?: string }
  warnings: string[]
  scope_note?: string
  unlinked_notes?: { title: string; text: string }[]
  diagnostics?: HighTrendCard[]
  has_content: boolean
  partial: boolean
}

export type HighTrendStatus = 'queued' | 'running' | 'completed' | 'partial' | 'empty' | 'failed'

/** 新结果只包含 LIKE / ENJOY；旧历史结果可能仍包含未标注附件。 */
export interface UserResearchImage {
  path: string
  file_exists: boolean
  user_ids: (string | number)[]
  image_ids: string[]
  emotion_tags?: string[]
  linked_to_result: boolean
  url?: string | null
}

export interface HighTrendTask {
  id: string
  status: HighTrendStatus
  stage: string
  message: string
  completed_jobs: number
  total_jobs: number
  created_at: string
  updated_at: string
  request: HighTrendRequest
  scope?: HighTrendScope
  counts?: HighTrendPreview['counts']
  error?: string | null
  error_detail?: { code: string; message: string; exception_type: string; job_id?: string | null; seconds?: number | null; http_status?: number | null; time: string }
  error_history?: HighTrendTask['error_detail'][]
  resume?: { can_resume: boolean; used_calls: number; remaining_jobs: number; minimum_max_calls: number }
  events: { time: string; stage: string; message: string }[]
  result?: HighTrendResult | null
  performance?: Record<string, unknown> | null
}

File: backend/app/schemas/high_trends.py
"""高潜趋势运行参数；日期只约束趋势资料，不筛选用户回答。"""

from datetime import date
from typing import Literal
from pydantic import BaseModel, Field, model_validator


class HighTrendRequest(BaseModel):
    start_date: date
    end_date: date
    model_id: str
    dataset: Literal["original", "article_table_2_selected_5"] = Field(default="original", description="使用已登记的数据组合，不接受任意文件路径")
    all_dates: bool = Field(default=False, description="纳入全部趋势，包括无日期数据；此时忽略日期区间")
    max_calls: int = Field(default=24, ge=1, le=100)
    prompt: str = Field(default="", max_length=4000, description="用户补充的设计关注方向和输出要求")
    user_scope: Literal["auto", "all", "first"] = Field(default="auto", description="自动识别文字范围、全部用户或前 N 位用户")
    user_limit: int | None = Field(default=None, ge=1, strict=True, description="按原始资料顺序选择的用户数，不是回答条数")

    @model_validator(mode="after")
    def validate_range(self):
        if not self.all_dates and self.start_date > self.end_date:
            raise ValueError("开始日期不能晚于结束日期")
        if self.user_scope == "first" and self.user_limit is None:
            raise ValueError("请选择前几位用户")
        if self.user_scope != "first" and self.user_limit is not None:
            raise ValueError("填写用户数量时，请选择前 N 位用户模式")
        return self


class HighTrendResumeRequest(BaseModel):
    """继续沿用原始范围和模型，仅允许显式调整累计调用预算。"""
    max_calls: int | None = Field(default=None, ge=1, le=100)
